formats/opentype.fathom

```diff
@@ -185,6 +186,197 @@ let language_id =
 let language_id32 =
     u32be;
 
+
+// # Common Table Formats
+
+/// # Class Definition Table
+///
+/// | Class | Description                                               |
+/// |-------|-----------------------------------------------------------|
+/// | 1     | Base glyph (single character, spacing glyph)              |
+/// | 2     | Ligature glyph (multiple character, spacing glyph)        |
+/// | 3     | Mark glyph (non-spacing combining glyph)                  |
+/// | 4     | Component glyph (part of single character, spacing glyph) |
+///
+/// ## References
+///
+/// - [Microsoft's OpenType Spec: Class Definition Table](https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#class-definition-table)
+
+/// # Class Definition Table Format 1
+///
+/// ## References
+///
+/// - [Microsoft's OpenType Spec: Class Definition Table Format 1](https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#class-definition-table-format-1)
+let class_def_format_1 = {
+    /// First glyph ID of the class_value_array
+    start_glyph_id <- u16be,
+    /// Size of the class_value_array
+    glyph_count <- u16be,
+    /// Array of Class Values — one per glyph ID
+    class_value_array <- array16 glyph_count u16be,
+};
+
+/// # Class Definition Table Format 2
+///
+/// ## References
+///
+/// - [Microsoft's OpenType Spec: Class Definition Table Format 2](https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#class-definition-table-format-2)
+let class_def_format_2 = (
+    /// ClassRangeRecord
+    let class_range_record = {
+        /// First glyph ID in the range
+        start_glyph_id <- u16be,
+        /// Last glyph ID in the range
+        end_glyph_id <- u16be,
+        /// Applied to all glyphs in the range
+        class <- u16be,
+    };
+
+    {
+        /// Number of ClassRangeRecords
+        class_range_count <- u16be,
+        /// Array of ClassRangeRecords — ordered by startGlyphID
+        class_range_records <- array16 class_range_count class_range_record,
+    }
+);
+
+/// # Class Definition Table
+///
+/// ## References
+///
+/// - [Microsoft's OpenType Spec: Class Definition Table](https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#class-definition-table)
+let class_def = {
+    /// Format identifier
+    class_format <- u16be,
+    /// Format specific data
+    data <- match class_format {
+        1 => class_def_format_1,
+        2 => class_def_format_2,
+        _ => fail,
```

Comment:
This is a bit of an open question in my mind, at least partially informed by the use case. I'm picturing two use-cases: exploring a font file, and parsing and using that data in a font file. I suppose I've been thinking more along the lines of the latter, which might actually be wrong for where we're at.

Where I've used `fail` in this PR there is no data to fall back on. For example, If someone was relying on getting class def information we'd be unable to provide it if it used a format we were unaware of. Presumably the alternative is return something in place of failing, do you have thoughts on what that could be?

In other places like the handling of the `GDEF` minor version we fall back on the newest version we know about as each of the revisions is additive so we can read up to the version we know.

Reply:
I think the way I've been handling it is with the `unknown_table` format which is just a empty record format that always succeeds. 

```
/// # Unknown table format
///
/// This is a placeholder for a table that has an unknown identifier (due to the
/// file conforming to a newer version of the specification), or for a table has
/// not yet been implemented.
let unknown_table : Format = {};
```

It used to be used as a fallback case to a match arm, before we switched to our current approach for font tables. This meant that the match arm would succeed if a new version was introduced that we didn't recognise, but no data would be read. Unfortunately this approach needs to be used with care… like you need to use it at the end of a substream, otherwise the format could continue to be read afterwards. Though I think this would only really occur in a poorly designed format I guess.

Reply:
Ahh, we currently use it for the cmap tables:

https://github.com/yeslogic/fathom/blob/811673a3833c66942e3282ad1e8c396c05d0ab1b/formats/opentype.fathom#L565-L583

Reply:
I’ve had a bit of a think about it, and don't think this is a huge issue right now, so feel free to ignore me. And yeah thanks for the consideration of the different use cases… seems that the desired behavior might be a bit application specific. Food for thought at the very least!

Reply:
Heh I pondered it too and decided to change it to `unknown_table`:

- Nothing else was using fail.
- The doc comments on `unknown_table` are clear that it's for cases like this.
- We can change it to something like `{ unknown <- {} }` as you proposed in the future to make it show up in the output if we want.